src/metabase/models/secret.clj

```diff
@@ -269,70 +116,270 @@
+        path (get details (:path kws))
+        path-map (when has-path?
+                   {:source :file-path :value path})
+        value-map (when has-value?
+                    {:source :uploaded :value value})
+        secret-map (case (keyword options)
+                     :local
+                     path-map
+
+                     :uploaded
+                     value-map
+
+                       ;; fallback
+                     (cond
+                       has-value? value-map
+                       has-path? path-map))]
+    (when (and path (premium-features/is-hosted?))
+      (throw (ex-info
+              (tru "{0} (a local file path) cannot be used in Metabase hosted environment" (:path kws))
+              {:invalid-db-details-entry (select-keys details [(:path kws)])})))
+
+    ;; If the client sent us back protected-password then it should be ignored and value loaded from Secret.
+    (when (and secret-map (not= (:value secret-map) protected-password))
+      (update secret-map :value #(some-> % codecs/to-bytes)))))
+
+(defn- resolve-secret-map
+  "Returns a canonical map containing `:value` and `:source` for the given `secret-property`.
+   May also include `:id` if the value is coming from a persisted Secret.
+
+   If `:details` contains expanded secrets from the client (i.e. during connection testing) it will be preferred.
+   If a `-id` property exists, that will be used to lookup the Secret.
+   Otherwise return nil."
+  [driver details secret-property]
+  (let [conn-prop (get (secret-conn-props-by-name driver) secret-property)
+        detail-map (secret-map-from-details details conn-prop)
+        secret-id (get details (->id-kw secret-property))
+        result (cond
+                 detail-map detail-map
+                 secret-id (latest-for-id secret-id))
+        result-source (:source result)]
+    (when (:value result)
+      (cond-> result
+        ;; Normalizes legacy
+        (not result-source) (assoc :source :uploaded)))))
+
+(defn- unresolved-value-string
+  "Reads the secret-value, which is probably bytes into a string.
+
+   Private because getting the file-path is an implementation detail of Secret."
+  [secret-value]
+  (cond (string? secret-value)
+        secret-value
+        (bytes? secret-value)
+        (u/bytes-to-string secret-value)))
+
+;;; ---------------------------------------------- Fetching secrets ----------------------------------------------
+
+(defn value-as-string
+  "Retrieves a secret as a string.
+   If the secret source is `:file-path` then read the file and return the contents.
+   Otherwise return the secret value."
+  [driver details secret-property]
+  (when-let [{source :source secret-value :value} (resolve-secret-map driver details secret-property)]
+    (let [s (unresolved-value-string secret-value)]
+      (if (= :file-path source)
+        (slurp s)
+        s))))
+
+(defn- value-as-file*
+  [driver details secret-property & [ext]]
```

Comment:
I prefer two separate arities as opposed to allowing any number of arguments, so that linters can check the calls.